modules/3d/src/rgbd/odometry_frame_impl.cpp

```diff
@@ -0,0 +1,251 @@
+{
+};
+
+template<typename TMat>
+void OdometryFrameImplTMat<TMat>::setImage(InputArray _image)
+{
+    this->image = getTMat<TMat>(_image);
+    Mat gray;
+    if (_image.channels() != 1)
+        cvtColor(_image, gray, COLOR_BGR2GRAY, 1);
+    else
+        gray = getTMat<Mat>(_image);
+    gray.convertTo(gray, CV_8UC1);
+    this->imageGray = getTMat<TMat>(gray);
+}
+
+template<typename TMat>
+void OdometryFrameImplTMat<TMat>::getImage(OutputArray _image) const
+{
+    _image.assign(this->image);
+}
+
+template<typename TMat>
+void OdometryFrameImplTMat<TMat>::getGrayImage(OutputArray _image) const
+{
+    _image.assign(this->imageGray);
+}
+
+template<typename TMat>
+void OdometryFrameImplTMat<TMat>::setDepth(InputArray _depth)
+{
+
+    TMat depth_tmp, depth_flt;
+    depth_tmp = getTMat<TMat>(_depth);
+    // Odometry works with depth values in range [0, 10)
+    // if the max depth value more than 10, it needs to devide by 5000
+    double max;
+    cv::minMaxLoc(depth_tmp, NULL, &max);
+    if (max > 10)
+    {
+        depth_tmp.convertTo(depth_flt, CV_32FC1, 1.f / 5000.f);
+        depth_flt.setTo(std::numeric_limits<float>::quiet_NaN(), getTMat<Mat>(depth_flt) < FLT_EPSILON);
+        depth_tmp = depth_flt;
+    }
+    this->depth = depth_tmp;
+    this->findMask(_depth);
+}
+
+template<typename TMat>
+void OdometryFrameImplTMat<TMat>::getDepth(OutputArray _depth) const
+{
+    _depth.assign(this->depth);
+}
+
+template<typename TMat>
+void OdometryFrameImplTMat<TMat>::setMask(InputArray _mask)
+{
+    this->mask = getTMat<TMat>(_mask);
+}
+
+template<typename TMat>
+void OdometryFrameImplTMat<TMat>::getMask(OutputArray _mask) const
+{
+    _mask.assign(this->mask);
+}
+
+template<typename TMat>
+void OdometryFrameImplTMat<TMat>::setNormals(InputArray _normals)
+{
+    this->normals = getTMat<TMat>(_normals);
+}
+
+template<typename TMat>
+void OdometryFrameImplTMat<TMat>::getNormals(OutputArray _normals) const
+{
+    _normals.assign(this->normals);
+}
+
+template<typename TMat>
+void OdometryFrameImplTMat<TMat>::setPyramidLevel(size_t _nLevels, OdometryFramePyramidType oftype)
+{
+    if (oftype < OdometryFramePyramidType::N_PYRAMIDS)
+        pyramids[oftype].resize(_nLevels, TMat());
+    else
+        CV_Error(Error::StsBadArg, "Incorrect type.");
+
+}
+
+template<typename TMat>
+void OdometryFrameImplTMat<TMat>::setPyramidLevels(size_t _nLevels)
+{
+    for (auto& p : pyramids)
+    {
+        p.resize(_nLevels, TMat());
+    }
+}
+
+template<typename TMat>
+size_t OdometryFrameImplTMat<TMat>::getPyramidLevels(OdometryFramePyramidType oftype) const
+{
+    if (oftype < OdometryFramePyramidType::N_PYRAMIDS)
+        return pyramids[oftype].size();
+    else
+        return 0;
+}
+
+template<typename TMat>
+void OdometryFrameImplTMat<TMat>::setPyramidAt(InputArray  _img, OdometryFramePyramidType pyrType, size_t level)
+{
+    CV_Assert(pyrType >= 0);
+    CV_Assert(pyrType < pyramids.size());
```

Comment:
When our branch merged this part of the code, the build gave a WARNING and the logs are at https://pullrequest.opencv.org/buildbot/builders/precommit_windows32/builds/13578/steps/compile%20release/logs/warnings%20%281%29.

Reply:
`CV_Assert((size_t)pyrType < pyramids.size());`

Reply:
@alalek, I'll fix it on this PR or create a new one?

Reply:
Nobody can fix already merged PR.
So create a new one.

Reply:
[PR#21272](https://github.com/opencv/opencv/pull/21272)